### PR TITLE
Don't select random beatmap when previous was filtered

### PR DESCRIPTION
**Pull request by LittleEndu**
* Supersedes #8384 (deleted my previous fork so can't push there)

This PR will not let carousel root use a random selection when previous selection is filtered and instead makes it use the default selection (the default is `GetNextToSelect()`). I think this is preferred, because when the user doesn't explicitly randomize the carousel then the selection should be near the previous one (so the user wouldn't need to scroll back to where their previous selection was after carousel selected a random beatmap)

The diff for tests looks weird, but I'm actually changing out the test completely.

As per [peppy's comment](https://github.com/ppy/osu/pull/8384#discussion_r547059703) readability of `GetNextToSelect()` has been improved. The code that figures out the distance to next beatmap now only counts filtered beatmaps, but that means using `.SkipWhile(pred + counting)` instead of `.FirstOrDefault(pred)` which sounds like a performance concern. Although we can afford `GetNextToSelect()` being a bit slow to perform because carousel only ever calls it eagerly when it loses its selection (which can only happen during filter changes).

### CODE REVIEW

File: osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapCarousel.cs [hunk 1]
```diff
@@ -616,55 +616,72 @@ public void TestCarouselRemembersSelection()
             AddAssert("Selection was remembered", () => eagerSelectedIDs.Count == 1);
         }
 
-        [Test]
-        public void TestRandomFallbackOnNonMatchingPrevious()
+        [TestCase(10, 1)]
+        [TestCase(10, 10)]
+        public void TestCarouselSelectsNextWhenPreviousIsFiltered(int makeThisManyGroups, int haveThisManySetsInGroup)
```
Comment (peppy) [inline osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapCarousel.cs:621]: This isn't how we name variables

`groupCount` and `setsPerGroup` would work better.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
I'm still waiting on more feedback from others regarding whether this is the UX we want, but I've also noticed that the tests don't even cover the new behaviour you've added. You'd want to add a test showing how your change "remembers" previous search history in a potentially unexpected way.

Are you waiting for feedback before adding test coverage or is this an oversigth?

**Comment by LittleEndu** [comment]
Just to make sure peppy, are you commenting on the right PR? I have two PR changing carousel behaviour open, this is the one which changes behaviour when applying a new filter, the other (https://github.com/ppy/osu/pull/14572) is the one changing behaviour when returning to old filters.

Pretty sure my tests from 2020 are testing the new behaviour of carousel root not selecting randomly anymore. But I wrote them in 2020 so I don't remember.

**Comment by peppy** [comment]
@LittleEndu I still generally agree with the premise of this change, but the tests are impossible to understand. I've brought the code to a compiling state, but the tests don't work.

If you're still interested, please rewrite the tests with comments (and required fixes) and either reopen this PR or open a fresh one.

**Pull request closed without merging**